File: src/main/java/io/zeitwert/fm/portfolio/model/base/ObjPortfolioBase.java

package io.zeitwert.fm.portfolio.model.base;

import static io.dddrive.util.Invariant.requireThis;

import java.math.BigDecimal;
import java.security.InvalidParameterException;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import io.dddrive.ddd.model.enums.CodeAggregateType;
import io.dddrive.ddd.model.enums.CodeAggregateTypeEnum;
import io.dddrive.obj.model.Obj;
import io.dddrive.property.model.ReferenceSetProperty;
import io.dddrive.property.model.SimpleProperty;
import io.zeitwert.fm.account.model.ObjAccount;
import io.zeitwert.fm.building.model.ObjBuilding;
import io.zeitwert.fm.building.model.db.tables.records.ObjBuildingVRecord;
import io.zeitwert.fm.collaboration.model.impl.AggregateWithNotesMixin;
import io.zeitwert.fm.obj.model.base.FMObjBase;
import io.zeitwert.fm.obj.service.api.ObjVCache;
import io.zeitwert.fm.oe.model.ObjTenantFM;
import io.zeitwert.fm.portfolio.model.ObjPortfolio;
import io.zeitwert.fm.portfolio.model.ObjPortfolioRepository;
import io.zeitwert.fm.task.model.impl.AggregateWithTasksMixin;

public abstract class ObjPortfolioBase extends FMObjBase
		implements ObjPortfolio, AggregateWithNotesMixin, AggregateWithTasksMixin {

	private static final List<CodeAggregateType> OBJ_TYPES = List.of(
			CodeAggregateTypeEnum.getAggregateType("obj_portfolio"),
			CodeAggregateTypeEnum.getAggregateType("obj_account"),
			CodeAggregateTypeEnum.getAggregateType("obj_building"));

	//@formatter:off
	protected final SimpleProperty<String> name = this.addSimpleProperty("name", String.class);
	protected final SimpleProperty<String> description = this.addSimpleProperty("description", String.class);
	protected final SimpleProperty<String> portfolioNr = this.addSimpleProperty("portfolioNr", String.class);
	protected final ReferenceSetProperty<Obj> includeSet = this.addReferenceSetProperty("includeSet", Obj.class);
	protected final ReferenceSetProperty<Obj> excludeSet = this.addReferenceSetProperty("excludeSet", Obj.class);
	protected final ReferenceSetProperty<ObjBuilding> buildingSet = this.addReferenceSetProperty("buildingSet", ObjBuilding.class);
	//@formatter:on

	protected ObjPortfolioBase(ObjPortfolioRepository repository, Object state) {
		super(repository, state);
	}

	@Override
	public ObjPortfolioRepository getRepository() {
		return (ObjPortfolioRepository) super.getRepository();
	}

	@Override
	public ObjPortfolio aggregate() {
		return this;
	}

	@Override
	public final ObjAccount getAccount() {
		return this.getRepository().getAccountCache().get(this.getAccountId());
	}

	public void addInclude(Integer id) {
		requireThis(this.hasValidObjType(id), "supported objType " + id);
		this.includeSet.addItem(id);
	}

	public void addExclude(Integer id) {
		requireThis(this.hasValidObjType(id), "supported objType " + id);
		this.excludeSet.addItem(id);
	}

	protected boolean hasValidObjType(Integer id) {
		Obj obj = this.getRepository().getObjCache().get(id);
		CodeAggregateType objType = obj.getMeta().getAggregateType();
		return OBJ_TYPES.indexOf(objType) >= 0;
	}

	@Override
	public double getInflationRate() {
		BigDecimal inflationRate = this.getAccount().getInflationRate();
		inflationRate = inflationRate != null ? inflationRate : ((ObjTenantFM) this.getTenant()).getInflationRate();
		return inflationRate != null ? inflationRate.doubleValue() : 0;
	}

	@Override
	public double getPortfolioValue(int year) {
		// if (this.getInsuredValueYear() != null && this.getInsuredValue() != null) {
		// return ObjBuildingBase.DefaultPriceIndex.priceAt(this.getInsuredValueYear(),
		// 1000.0 * this.getInsuredValue().doubleValue(), year,
		// this.getInflationRate());
		// }
		return 0;
	}

	@Override
	public Integer getCondition(int year) {
		// ObjBuildingPartRating rating = this.getCurrentRating();
		// if (rating != null) {
		// return rating.getCondition(year);
		// }
		return null;
	}

	@Override
	protected void doCalcAll() {
		super.doCalcAll();
		this.calcCaption();
		this.calcBuildingSet();
	}

	private void calcCaption() {
		this.setCaption(this.getName());
	}

	private void calcBuildingSet() {
		this.buildingSet.clearItems();
		for (Integer objId : this.includeSet.getItems()) {
			this.getBuildingIds(objId).forEach(id -> this.buildingSet.addItem(id));
		}
		for (Integer objId : this.excludeSet.getItems()) {
			this.getBuildingIds(objId).forEach(id -> this.buildingSet.removeItem(id));
		}
	}

	private Set<Integer> getBuildingIds(Integer id) {
		ObjVCache objCache = this.getRepository().getObjCache();
		Obj obj = objCache.get(id);
		CodeAggregateType objType = obj.getMeta().getAggregateType();
		if (objType == CodeAggregateTypeEnum.getAggregateType("obj_building")) {
			return Set.of(obj.getId());
		} else if (objType == CodeAggregateTypeEnum.getAggregateType("obj_portfolio")) {
			ObjPortfolio pf = this.getRepository().getPortfolioCache().get(id);
			return pf.getBuildingSet();
		} else if (objType == CodeAggregateTypeEnum.getAggregateType("obj_account")) {
			List<ObjBuildingVRecord> buildings = this.getRepository().getBuildingRepo().getByForeignKey("account_id", id);
			return buildings.stream().map(bldg -> bldg.getId()).collect(Collectors.toSet());
		}
		throw new InvalidParameterException("unsupported objType " + objType.getId());
	}

	@Override
	public void doCalcSearch() {
		super.doCalcSearch();
		this.addSearchToken(this.getPortfolioNr());
		this.addSearchText(this.getName());
		this.addSearchText(this.getDescription());
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/portfolio/model/base/ObjPortfolioBase.kt
package io.zeitwert.fm.portfolio.model.base

import io.dddrive.ddd.model.enums.CodeAggregateType
import io.dddrive.ddd.model.enums.CodeAggregateTypeEnum
import io.dddrive.obj.model.Obj
import io.dddrive.property.model.ReferenceSetProperty
import io.zeitwert.fm.building.model.ObjBuilding
import io.zeitwert.fm.collaboration.model.impl.AggregateWithNotesMixin
import io.zeitwert.fm.obj.model.base.FMObjBase
import io.zeitwert.fm.oe.model.ObjTenantFM
import io.zeitwert.fm.portfolio.model.ObjPortfolio
import io.zeitwert.fm.portfolio.model.ObjPortfolioRepository
import io.zeitwert.fm.task.model.impl.AggregateWithTasksMixin

abstract class ObjPortfolioBase(
	override val repository: ObjPortfolioRepository,
	isNew: Boolean,
) : FMObjBase(repository, isNew),
	ObjPortfolio,
	AggregateWithNotesMixin,
	AggregateWithTasksMixin {

	companion object {

		private val OBJ_TYPES: List<CodeAggregateType> by lazy {
			listOf(
				CodeAggregateTypeEnum.getAggregateType("obj_portfolio"),
				CodeAggregateTypeEnum.getAggregateType("obj_account"),
				CodeAggregateTypeEnum.getAggregateType("obj_building"),
			)
		}
	}

	private lateinit var _includeSet: ReferenceSetProperty<Obj>
	private lateinit var _excludeSet: ReferenceSetProperty<Obj>
	private lateinit var _buildingSet: ReferenceSetProperty<ObjBuilding>

	override fun doInit() {
		super.doInit()
		addBaseProperty("name", String::class.java)
		addBaseProperty("description", String::class.java)
		addBaseProperty("portfolioNr", String::class.java)
		_includeSet = addReferenceSetProperty("includeSet", Obj::class.java)
		_excludeSet = addReferenceSetProperty("excludeSet", Obj::class.java)
		_buildingSet = addReferenceSetProperty("buildingSet", ObjBuilding::class.java)
	}

	override fun aggregate(): ObjPortfolio = this

	override fun taskRepository() = repository.taskRepository

	override val account get() = if (accountId != null) repository.accountRepository.get(accountId!!) else null

	// override fun addInclude(id: Int?) {
	// 	require(hasValidObjType(id)) { "supported objType $id" }
	// 	_includeSet.addItem(id)
	// }

	// override fun addExclude(id: Int?) {
	// 	requireThis(hasValidObjType(id), "supported objType $id")
	// 	_excludeSet.addItem(id)
	// }

	private fun hasValidObjType(id: Int?): Boolean {
		if (id == null) return false
		val obj = repository.get(id)
		val objType = obj?.meta?.repository?.aggregateType
		return OBJ_TYPES.contains(objType)
	}

	override val inflationRate: Double
		get() {
			val inflationRate = account?.inflationRate
				?: (tenant as? ObjTenantFM)?.inflationRate
			return inflationRate?.toDouble() ?: 0.0
		}

	override val discountRate: Double
		get() {
			val discountRate = account?.discountRate
				?: (tenant as? ObjTenantFM)?.discountRate
			return discountRate?.toDouble() ?: 0.0
		}

	override fun getPortfolioValue(year: Int): Double = 0.0

	override fun getCondition(year: Int): Int = 0

	override fun doCalcAll() {
		super.doCalcAll()
		calcCaption()
		calcBuildingSet()
	}

	private fun calcCaption() {
		setCaption(name)
	}

	private fun calcBuildingSet() {
		_buildingSet.clearItems()
		for (objId in _includeSet.items) {
			getBuildingIds(objId as Int).forEach { _buildingSet.addItem(it) }
		}
		for (objId in _excludeSet.items) {
			getBuildingIds(objId as Int).forEach { _buildingSet.removeItem(it) }
		}
	}

	private fun getBuildingIds(id: Any): Set<Any> {
		val objType = repository
			.get(id)
			.meta.repository.aggregateType
		return when (objType.id) {
			"obj_building" -> {
				setOf(id)
			}

			"obj_portfolio" -> {
				val pf = repository.get(id)
				pf.buildingSet
			}

			"obj_account" -> {
				repository
					.buildingRepository
					.getByForeignKey("accountId", id)
					.map { it as Int }
					.toSet()
			}

			else -> {
				emptySet()
			}
		}
	}

	// override fun doCalcSearch() {
	//     super.doCalcSearch()
	//     this.addSearchToken(this.portfolioNr)
	//     this.addSearchText(this.name)
	//     this.addSearchText(this.description)
	// }

}
